sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pxd

```diff
@@ -18,4 +20,33 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
     cdef:
         const DTYPE_t[:, ::1] X
         const DTYPE_t[:, ::1] Y
-        ITYPE_t d
+
+
+cdef class SparseSparseDatasetsPair(DatasetsPair):
+    cdef:
+        const DTYPE_t[:] X_data
+        const SPARSE_INDEX_TYPE_t[:] X_indices
+        const SPARSE_INDEX_TYPE_t[:] X_indptr
+
+        const DTYPE_t[:] Y_data
+        const SPARSE_INDEX_TYPE_t[:] Y_indices
+        const SPARSE_INDEX_TYPE_t[:] Y_indptr
```

Comment:
Is the plan to Tempita all of these? Specifically, `DTYPE_t` to become `float32` and `float64` and `SPARSE_INDEX_TYPE_t` to become `int32` and `int64`.

If so, does this extend to all combinations? For example, sparse data with `float32` data and `int64` indices paired with a sparse data with `float64` data and `int32` indices. 🤯

Reply:
Potentially yes, especially for `DTYPE_t` which will remain fixed datasetpair-wise.

I don't see how to easily handle various the case on `SPARSE_INDEX_TYPE_t` yet/using Cython, and I fear the complex logic and manual templating…


Other languages would have allowed to ease some logic, but going done this road come after broader discussions IMO.